opendevin/server/session/session.py

```diff
@@ -96,6 +96,11 @@ async def dispatch(self, data: dict):
             await self._initialize_agent(data)
             return
         event = event_from_dict(data.copy())
+        event._source = EventSource.USER  # type: ignore[attr-defined]
```

Comment:
Maybe only add this when the `event` does not contain `_source`?

We are working on a refractor where it is possible that the `od-runtime-cli` will also send event to the backend: https://github.com/OpenDevin/OpenDevin/issues/2404

In that case, the event source will likely NOT user.

Reply:
I'm a bit confused here. The commit that addresses this doesn't change it only for session.py, it changes the event stream too, changing the EventSource parameter to optional, thus allowing the possibility that it's not set when added to the stream.

@xingyaoww is that what you had in mind? Currently it's not optional. All client code uses `add_event(event, EventSource)` when it adds the event to the event stream. Once it was added, we _know_ it has been set. Will `od-runtime-cli` not know the source? I don't know why it wouldn't... In fact, your comment implies that `event` will contain `_source`, so can't we just add the event to the stream in session.py with that source when set, and fallback to USER when not set?

The relevant code is just below this line:
https://github.com/OpenDevin/OpenDevin/pull/2532/files/a2d91ee3a809fec2516b17bf135f18ec3753fc56#diff-41af609e38c9fefa79dbea822eea7f42c03c9024caf3b51f991a59f0169dec5eL99

Reply:
When event.source will have None values here?

https://github.com/OpenDevin/OpenDevin/blob/6879acdfe745453d336c3d8b9c597ab256d52db2/opendevin/runtime/runtime.py#L116

Reply:
@SmartManoj Always. `on_event` is the callback function for AgentController, which is called just after an event was added to the stream. When it was added to the stream, it got `_source` set. So it will always be non-None here.

Reply:
So safe to change to, `await self.event_stream.add_event(observation, EventSource.AGENT)`?

Reply:
Hmm. When I saw this code, I first thought it was in AgentController. It's in the runtime. It should be the same, it's the runtime's callback function, but there's a tweak: the runtime creates observations. I need to dig more to see why exactly this exact code was assuming that it might not have it set _here_.

Reply:
Can't we just do `event._source = EventSource.USER if not event._source else event._source` or something like that, at the line Xingyao mentioned?

Reply:
The file is outdated now.

Reply:
> changing the EventSource parameter to optional, thus allowing the possibility that it's not set when added to the stream.

I'd actually prefer making the "source" mandatory to make sure we did not mess things up..

Reply:
reverted this.